src/device/wg_controller.rs

```diff
@@ -2189,128 +2314,269 @@ mod tests {
     }
 
     #[tokio::test]
-    async fn when_cross_check_validated_is_newer_then_upgrade() {
+    async fn when_ep_is_validated_after_wg_update_produce_no_failed_notification() {
+        // No failed notification should be provided regardless
+        // whether we are proxying or in direct connection.
+        //
+        // Test for that both of these cases
+        let mapped_port = 12;
+        let proxy_endpoint = SocketAddr::from(([127, 0, 0, 1], mapped_port));
+        let local_wg_endpoint = SocketAddr::from(([192, 168, 0, 2], 15));
+        let wg_endpoints = vec![proxy_endpoint, local_wg_endpoint];
+
+        // Test all wg_endpoint cases
+        for wg_endpoint in wg_endpoints {
+            let mut f = Fixture::new();
+            f.features.ipv6 = true;
+
+            let pub_key = SecretKey::gen().public();
+            let allowed_ips = Fixture::create_any_allowed_ips();
+            let remote_wg_endpoint = SocketAddr::from(([192, 168, 0, 1], 13));
+            let changed_endpoint_at = Instant::now();
+            let cpc_endpoint_at = changed_endpoint_at + Duration::from_secs(1);
+
+            // Arrange proxy mappings
+            f.when_proxy_mapping(vec![(pub_key, mapped_port)]);
+
+            // Arrange peers in wireguard_interface
+            f.when_current_peers(vec![(
+                pub_key,
+                wg_endpoint,
+                TEST_PERSISTENT_KEEPALIVE_PERIOD,
+                allowed_ips.clone(),
+                (changed_endpoint_at, UpdateReason::Pull),
+            )]);
+
+            // Arrange a list of validated CPC endpoints
+            f.when_cross_check_validated_endpoints(vec![(
+                pub_key,
+                remote_wg_endpoint,
+                (local_wg_endpoint, Fixture::create_any_session_id()),
+                cpc_endpoint_at,
+            )]);
+
+            // Prepare assertions
+            f.then_cross_ping_check_not_notfied_about_failed_connections(vec![pub_key]);
+
+            // Act
+            f.consolidate_endpoint_invalidation().await;
+        }
+    }
+
+    #[tokio::test]
+    async fn when_ep_is_validated_before_downgrade_produce_failed_notification() {
+        // EndpointGone notification should be produced regardless wether we
```

Comment:
Incredibly minor nit, only fix if fixing something else:
```suggestion
        // EndpointGone notification should be produced regardless whether we
```